### PR TITLE
add new json functions

### PR DESCRIPTION
**Pull request by yaojiefeng**

### CODE REVIEW

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 1]
```diff
@@ -197,12 +218,121 @@ public static JsonElement parse(String json) {
...
     return element;
   }
+
+  /**
+   * Parses a column or string to JSON. This is equivalent to <code>JSON.parse()</code>
+   *
+   * @param json string representation of json.
+   * @return parsed json else throws an exception.
+   */
+  public static JsonElement Parse(String json) {
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:236]: this behaves differently than the old parse() implementation. Is that intended?

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 2]
```diff
@@ -197,12 +218,121 @@ public static JsonElement parse(String json) {
...
+    try {
+      JsonElement element = PARSER.parse(json);
+      return element;
+    } catch (JsonSyntaxException e) {
+      return JsonNull.INSTANCE;
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:241]: is this the intended behavior? Seems odd to return null for an invalid value.

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 3]
```diff
@@ -197,12 +218,121 @@ public static JsonElement parse(String json) {
...
+   * Checks if a Json is {@code JsonNull}.
+   *
+   * @param element to be inspected for null.
+   * @return true if null, false otherwise.
+   */
+  public static boolean IsNull(JsonElement element) {
+    if (element != null && element.isJsonNull()) {
+      return true;
+    }
+    return false;
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: if the element itself is a null this returns false? Is that intended?

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java [hunk 4]
```diff
@@ -197,12 +218,121 @@ public static JsonElement parse(String json) {
...
+    return element != null && element.isJsonArray();
+  }
+
+  /**
+   * Selects part of JSON using JSON DSL specified as json path.
+   *
+   * @param element json to be inspected.
+   * @param path to be searched for in the element.
+   * @param paths other paths.
+   * @return A json element containing the results of all json paths.
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: json array?

File: wrangler-core/src/test/java/io/cdap/functions/JsonFunctionsTest.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    String[] directives = new String[]{
+      "set-column mayo json:Parse(body)",
+      "drop boody",
+      "set-column entries json:Select(mayo, '$.list[*].set.*')"
+    };
+    List<Row> rows = Arrays.asList(
+      new Row("body", JSON_SELECTION_EG1)
+    );
+    rows = TestingRig.execute(directives, rows);
+    Assert.assertTrue(rows.size() == 1);
```
Comment (albertshau) [inline wrangler-core/src/test/java/io/cdap/functions/JsonFunctionsTest.java:76]: The directives could be doing all sorts of wrong things and this test would still pass. It should be checking the output.

File: wrangler-docs/functions/json.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    {
+      a : 1,
+      b : 2.0,
+      c : test
+      d : true
+    }
+```
+
+## Parse
+Parses a json into `json-element`. If there are any issues in parsing, the function would return
```
Comment (albertshau) [inline wrangler-docs/functions/json.md]: would return -> returns

Comment (albertshau) [reply wrangler-docs/functions/json.md]: same for other descriptions.

File: wrangler-docs/functions/json.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+if `json` has the input json specified, then the result of the operation would return `true`.
+
+```
+  set-column parsedjson json:Parse(json)
+  set-column object json:IsObject(parsedjson)
+```
+
+## Select
+Returns part of json specified by json path.
```
Comment (albertshau) [inline wrangler-docs/functions/json.md]: Is json path a standard? I would not know how to use this from the doc.

File: wrangler-docs/functions/json.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+### Input
+`json-element`
+
+### Output
+boolean(`true`|`false`)
+
+### Example
+
+if `json` has the input json specified, then the result of the operation would return `true`.
```
Comment (albertshau) [inline wrangler-docs/functions/json.md]: should be consistent with the examples. Some have 'json' as the field, others have 'body'. 'body' seems slightly better, as the term 'json' is mentioned a bunch of times.

**Comment by albertshau** [comment]
I'm concerned about adding two copies of what look the same functions (parse vs Parse), except they behave differently at edge conditions.

I also don't know why it's important to change things to upper case. It looks like there are a bunch of existing lowercase functions, why are we making things inconsistent?

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 1 above)
Comment (yaojiefeng) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:236]: I think the new implementation will not throw exception if the input is invalid json, instead it treats it as a json null, I think that is intended, though I am unsure if it is correct to do this

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 2 above)
Comment (yaojiefeng) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java:241]: Same above

File: wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java (hunk 3 above)
Comment (yaojiefeng) [reply wrangler-core/src/main/java/io/cdap/functions/JsonFunctions.java]: Fixed

File: wrangler-docs/functions/json.md (hunk 8 above)
Comment (yaojiefeng) [reply wrangler-docs/functions/json.md]: Fixed

File: wrangler-docs/functions/json.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+## Parse
+Parses a json into `json-element`. If there are any issues in parsing, the function returns
+`json-null`
+
+### Namespace
+`json`
+
+### Input
+`json-string`
```
Comment (albertshau) [inline wrangler-docs/functions/json.md]: the convention here looks different than in other PRs. Please ensure they are consistent

File: wrangler-docs/functions/json.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+### Input
+`json-element`
+
+### Output
+boolean(`true`|`false`)
+
+### Example
+
+if `body` has the input json specified, then the result of the operation returns `false`
```
Comment (albertshau) [inline wrangler-docs/functions/json.md]: if -> If

**Review by albertshau — APPROVED** [review]

**Pull request merged**